File: masar_mw/masar_mw/page/workflow_summary/workflow_summary.py
import frappe

@frappe.whitelist()
def get_count():
    return frappe.db.sql(f"""SELECT 
    COALESCE(tsa.status, 'Draft') AS status,
    COUNT(CASE WHEN tsa.status = 'Draft' THEN 1 END) AS draft_count,
    COUNT(CASE WHEN tsa.status = 'Initiation of Discussion' THEN 1 END) AS io_discussion_count,
    COUNT(CASE WHEN tsa.status = 'Discussion (Acquisition)' THEN 1 END) AS d_acquisition_count,
    COUNT(tsa.name)  AS docname_count,
    COUNT(CASE WHEN tsa.status = 'Under Operator Approval' THEN 1 END) AS uo_approved_count,
    COUNT(CASE WHEN tsa.status = 'Under SP Approval' THEN 1 END) AS up_approved_count,
    COUNT(CASE WHEN tsa.status = 'Approved' THEN 1 END) AS approved_count
    
FROM `tabSales Acquisition` tsa
WHERE tsa.status IN ('Draft', 'Initiation of Discussion', 'Discussion (Acquisition)', 'Under Operator Approval', 'Under SP Approval', 'Approved')
GROUP BY tsa.status AND tsa.name;""", as_dict =True)
    
@frappe.whitelist()
def duration_chart_data():
    results =  frappe.db.sql(f"""
     SELECT 
    twh.wf_from,
    SUM(twh.duration) AS total_duration
FROM `tabWorkflow History` twh
WHERE twh.duration IS NOT NULL
GROUP BY twh.wf_from 
ORDER BY CASE 
    WHEN twh.wf_from = 'Draft' THEN 1
    WHEN twh.wf_from = 'Initiation of Discussion' THEN 2
    WHEN twh.wf_from = 'Discussion (Acquisition)' THEN 3
    WHEN twh.wf_from = 'Under Operator Approval' THEN 4
    WHEN twh.wf_from = 'Under SP Approval' THEN 5
    ELSE 6
END;
    """, as_dict =True)
    wf_from = list()
    durations_seconds= list()
    for result in results:
        wf_from.append(result.get('wf_from'))
        durations_seconds.append(float(result.get('total_duration')))
    durations_hours = [duration / 3600 for duration in durations_seconds]
    durations_hours_rounded = [round(duration, 3) for duration in durations_hours]
    return {
        'wf_from' : wf_from , 
        'durations_hours_rounded': durations_hours_rounded
    }
    



@frappe.whitelist()
def get_duration_average():
    results =  frappe.db.sql(f"""
    SELECT 
        twh.wf_from,
        AVG (twh.duration) AS avg_duration
        FROM `tabWorkflow History` twh
        GROUP BY  twh.wf_from
    ORDER BY CASE 
        WHEN twh.wf_from = 'Draft' THEN 1
        WHEN twh.wf_from = 'Initiation of Discussion' THEN 2
        WHEN twh.wf_from = 'Discussion (Acquisition)' THEN 3
        WHEN twh.wf_from = 'Under Operator Approval' THEN 4
        WHEN twh.wf_from = 'Under SP Approval' THEN 5
        ELSE 6
    END;
    """, as_dict =True)
    wf_from = list()
    duration_avg = list()
    for result in results:
        # if result.get('wf_from') is None:
        wf_from.append(result.get('wf_from'))
        if result.get('avg_duration'):
            duration_avg.append(result.get('avg_duration'))
        else:
            duration_avg.append(0.0) 
    duration_avg_hours = [ duration / 3600 for duration in  duration_avg ]
    duration_avg_hours_rounded = [round(duration, 3) for duration in duration_avg_hours]
    return {
        'wf_from' : wf_from, 
        'duration_avg_hours_rounded' : duration_avg_hours_rounded
    }

@frappe.whitelist()
def status_percentage_data():
    result = frappe.db.sql("""
        SELECT 
            COALESCE(tsa.status, 'Draft') AS status,
            COUNT(CASE WHEN tsa.status = 'Draft' THEN 1 END) AS draft_count,
            COUNT(CASE WHEN tsa.status = 'Initiation of Discussion' THEN 1 END) AS io_discussion_count,
            COUNT(CASE WHEN tsa.status = 'Discussion (Acquisition)' THEN 1 END) AS d_acquisition_count,
            COUNT(CASE WHEN tsa.status = 'Under Operator Approval' THEN 1 END) AS uo_approved_count,
            COUNT(CASE WHEN tsa.status = 'Under SP Approval' THEN 1 END) AS up_approved_count,
            COUNT(CASE WHEN tsa.status = 'Approved' THEN 1 END) AS approved_count,
            COUNT(tsa.name)  AS docname_count
        FROM 
            `tabSales Acquisition` tsa
        WHERE 
            tsa.status IN ('Draft', 'Initiation of Discussion', 'Discussion (Acquisition)', 'Under Operator Approval', 'Under SP Approval', 'Approved')
        GROUP BY 
            tsa.status AND tsa.name
    """, as_dict = True)
    status_percentage = list()
    draft_percentage = float(result[0]['draft_count']) / float(result[0]['docname_count'])
    descussion_percentage = float( result[0]['io_discussion_count']) / float(result[0]['docname_count'])
    acquisition_percentage =  float(result[0]['d_acquisition_count']) / float(result[0]['docname_count'])
    uo_approved_percentage  =  float(result[0]['uo_approved_count']) / float(result[0]['docname_count'])
    up_approved_percentage  =  float(result[0]['up_approved_count']) / float(result[0]['docname_count'])
    approved_percentage  =  float(result[0]['approved_count']) / float(result[0]['docname_count'])
    status_percentage.append(round(draft_percentage ,3))
    status_percentage.append(round(descussion_percentage,3))
    status_percentage.append(round(acquisition_percentage ,3))
    status_percentage.append(round(uo_approved_percentage,3))
    status_percentage.append(round(up_approved_percentage ,3))
    status_percentage.append(round(approved_percentage,3))

    return {
        'status' : ['Draft', 'Initiation of Discussion', 'Discussion (Acquisition)', 'Under Operator Approval', 'Under SP Approval', 'Approved'], 
        'percentages' : status_percentage
    }

File: masar_mw/masar_mw/page/workflow_summary/workflow_summary.js

frappe.pages['workflow-summary'].on_page_load = function (wrapper) {
	new MyPage(wrapper);
}
MyPage = Class.extend({
	init: function (wrapper) {
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: 'Workflow Summary',
			single_column: true
		});
		this.make();
	},
	make: function () {
		var draft_count = 0;
		var io_discussion_count = 0;
		var d_acquisition_count = 0;
		var uo_approved_count = 0;
		var up_approved_count = 0;
		var approved_count = 0;
		var docname_count = 0;
		let me = $(this);
		let card_data = function(){
			frappe.call({
				method: "masar_mw.masar_mw.page.workflow_summary.workflow_summary.get_count",
				callback: function (r) {
					$.each(r.message, function (i, d) {
						draft_count = parseFloat(d.draft_count.toFixed());
						io_discussion_count = parseFloat(d.io_discussion_count.toFixed());
						d_acquisition_count = parseFloat(d.d_acquisition_count.toFixed(3));
						docname_count = parseFloat(d.docname_count.toFixed(3));
						uo_approved_count = parseFloat(d.uo_approved_count.toFixed(3));
						up_approved_count = parseFloat(d.up_approved_count.toFixed(3));
						approved_count = parseFloat(d.approved_count.toFixed(3));
					});
					$("#draft_count")[0].innerText = draft_count.toFixed(3)
					$("#io_discussion_count")[0].innerText = io_discussion_count.toFixed(3)
					$("#d_acquisition_count")[0].innerText = d_acquisition_count.toFixed(3)
					$("#docname_count")[0].innerText = docname_count.toFixed(3)
					$("#uo_approved_count")[0].innerText = uo_approved_count.toFixed(3)
					$("#up_approved_count")[0].innerText = up_approved_count.toFixed(3)
					$("#approved_count")[0].innerText = approved_count.toFixed(3)

				}
			});
		}
		let duration_chart = function() {
			frappe.call({
				method: 'masar_mw.masar_mw.page.workflow_summary.workflow_summary.duration_chart_data',
				callback: function(r) {
					let wf_from = r.message.wf_from;
					let duration = r.message.durations_hours_rounded;
					const data = {
						labels: wf_from,
						datasets: [
							{
								name: 'Duration Hours',
								values: duration
							}
						]
					};
					const chart = new frappe.Chart("#duration_chart", {  
								title: "Workflow History Chart",
								data: data,
								type: 'bar',
								height: 300,
								// colors: ['green'],
							})
				
				}
			});
		}

		let duration_avg_chart = function() {
			frappe.call({
				method: 'masar_mw.masar_mw.page.workflow_summary.workflow_summary.get_duration_average',
				callback: function(r) {
					let wf_from = r.message.wf_from;
					let duration_avg = r.message.duration_avg_hours_rounded;
					const data = {
						labels: wf_from,
						datasets: [
							{
								name: 'History Average',
								values: duration_avg
							}
						]
					};
					const chart = new frappe.Chart("#duration_avg_chart", {  
								title: "Stages Percentage",
								data: data,
								type: 'percentage',
								height: 300,
								// colors: ['#7cd6fd', '#743ee2'],
							})
				
				}
			});
		}
		let status_percentage_chart = function() {
			frappe.call({
				method: 'masar_mw.masar_mw.page.workflow_summary.workflow_summary.status_percentage_data',
				callback: function(r) {
					let status = r.message.status;
					let percentages = r.message.percentages;
					const data = {
						labels: status,
						datasets: [
							{
								name: 'Status Percentage',
								values: percentages
							}
						]
					};
					const chart = new frappe.Chart("#status_per_chart", {  
								title: " ",
								data: data,
								type: 'pie',
								height: 400,
								width:200,
								// colors: ['#7cd6fd', '#743ee2'],
							})
				
				}
			});
		}
		$(frappe.render_template(frappe.saving_funds.body, this)).appendTo(this.page.main)
		card_data();
		duration_chart();
		duration_avg_chart();
		status_percentage_chart();
	}
})


let body = '<script src="https://unpkg.com/frappe-charts@latest"></script>'
body += '<div class="widget-group ">'
body += '			<div class="widget-group-head">'
body += '				<div class="widget-group-control"></div>'
body += '			</div>'
body += '			<div class="widget-group-body grid-col-3">'

body += '				<div class="widget widget-shadow number-widget-box" data-widget-name="draft_count">'
body += '					<div class="widget-head">'
body += '					<div class="widget-label">'
body += '						<div class="widget-title">'
body += '							<div class="number-label">Draft</div>'
body += '						</div>'
body += '						<div class="widget-subtitle"></div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-body">'
body += '					<div class="widget-content">'
body += '						<div class="number" style="color:undefined" id="draft_count"> {draft_count} </div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-footer"></div>'
body += '			</div>'

body += '				<div class="widget widget-shadow number-widget-box" data-widget-name="io_discussion_count">'
body += '					<div class="widget-head">'
body += '					<div class="widget-label">'
body += '						<div class="widget-title">'
body += '							<div class="number-label">Initiation of Discussion</div>'
body += '						</div>'
body += '						<div class="widget-subtitle"></div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-body">'
body += '					<div class="widget-content">'
body += '						<div class="number" style="color:undefined" id="io_discussion_count"> {io_discussion_count} </div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-footer"></div>'
body += '			</div>'

body += '				<div class="widget widget-shadow number-widget-box" data-widget-name="d_acquisition_count">'
body += '					<div class="widget-head">'
body += '					<div class="widget-label">'
body += '						<div class="widget-title">'
body += '							<div class="number-label">Discussion (Acquisition)</div>'
body += '						</div>'
body += '						<div class="widget-subtitle"></div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-body">'
body += '					<div class="widget-content">'
body += '						<div class="number" style="color:undefined" id="d_acquisition_count"> {d_acquisition_count} </div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-footer"></div>'
body += '			</div>'

body += '				<div class="widget widget-shadow number-widget-box" data-widget-name="docname_count">'
body += '					<div class="widget-head">'
body += '					<div class="widget-label">'
body += '						<div class="widget-title">'
body += '							<div class="number-label">Total Document</div>'
body += '						</div>'
body += '						<div class="widget-subtitle"></div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-body">'
body += '					<div class="widget-content">'
body += '						<div class="number text-danger" style="color:undefined" id="docname_count"> {docname_count} </div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-footer"></div>'
body += '			</div>'


body += '				<div class="widget widget-shadow number-widget-box" data-widget-name="uo_approved_count">'
body += '					<div class="widget-head">'
body += '					<div class="widget-label">'
body += '						<div class="widget-title">'
body += '							<div class="number-label">Under Operator Approval</div>'
body += '						</div>'
body += '						<div class="widget-subtitle"></div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-body">'
body += '					<div class="widget-content">'
body += '						<div class="number" style="color:undefined" id="uo_approved_count"> {uo_approved_count} </div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-footer"></div>'
body += '			</div>'



body += '				<div class="widget widget-shadow number-widget-box" data-widget-name="up_approved_count">'
body += '					<div class="widget-head">'
body += '					<div class="widget-label">'
body += '						<div class="widget-title">'
body += '							<div class="number-label">Under SP Approval</div>'
body += '						</div>'
body += '						<div class="widget-subtitle"></div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-body">'
body += '					<div class="widget-content">'
body += '						<div class="number" style="color:undefined" id="up_approved_count"> {up_approved_count} </div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-footer"></div>'
body += '			</div>'

body += '				<div class="widget widget-shadow number-widget-box" data-widget-name="approved_count">'
body += '					<div class="widget-head">'
body += '					<div class="widget-label">'
body += '						<div class="widget-title">'
body += '							<div class="number-label">Approved</div>'
body += '						</div>'
body += '						<div class="widget-subtitle"></div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-body">'
body += '					<div class="widget-content">'
body += '						<div class="number" style="color:undefined" id="approved_count"> {approved_count} </div>'
body += '					</div>'
body += '				</div>'
body += '				<div class="widget-footer"></div>'
body += '			</div>'
body += '		</div>'
body += '	</div>' 
body += '<div id="duration_chart"></div>'
body += '<div id="line-chart"></div>'
body += '<div id="duration_avg_chart"></div>'
body += '<div id="line-chart"></div>'
body += '<div id="status_per_chart"></div>'
body += '<div id="line-chart"></div>'



frappe.saving_funds = {
	body: body
}


frappe.dom.set_style(`
:root {
  --charts-label-color: #f4f5f6;
  --charts-axis-line-color: #f4f5f6;
  --charts-tooltip-title: var(--charts-label-color);
  --charts-tooltip-label: var(--charts-label-color);
  --charts-tooltip-value: yellow;
  --charts-tooltip-bg: #f4f5f6;
  --charts-dataset-circle-stroke: #f4f5f6;
  --charts-legend-label: #959da5;
  --charts-legend-value: var(--charts-label-color);
}

.graph-svg-tip{background:#959da5;}


.graph-svg-tip.comparison li .tooltip-legend {
  width: 12px;
  margin-right: 8px;
}
`);